### PR TITLE
Add cut cone gizmo

### PR DESCRIPTION
**Pull request by DiogoMendonc-a**

### CODE REVIEW

File: engine/src/cubos/engine/gizmos/renderer.cpp [hunk 1]
```diff
@@ -90,3 +91,67 @@ void GizmosRenderer::initBoxPrimitive()
...
+
+    // Base B
+    for (int i = 16; i < 30; i += 2)
+    {
+        indices[iIndex++] = 16;
+        indices[iIndex++] = i + 1;
+        indices[iIndex++] = i + 2;
+        indices[iIndex++] = 16;
+        indices[iIndex++] = i + 2;
+        indices[iIndex++] = i + 3;
```
[the remark this answers is not in the document]
Comment (DiogoMendonc-a) [reply engine/src/cubos/engine/gizmos/renderer.cpp:154]: i hate you

File: engine/include/cubos/engine/gizmos/gizmos.hpp [hunk 2]
```diff
@@ -39,7 +39,7 @@ namespace cubos::engine
         /// @param from One of the ends of the line to be drawn.
         /// @param to The other end of the line to be drawn.
         /// @param lifespan How long the line will be on screen for, in seconds. Defaults to 0, which means a single
-        /// @param space Space to draw the gizmo in
+        /// @param space Space to draw the gizmo in.
         /// frame.
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gizmos/gizmos.hpp:43]:
```suggestion
        /// frame.
        /// @param space Space to draw the gizmo in.
```

File: engine/src/cubos/engine/gizmos/gizmos.cpp [hunk 3]
```diff
@@ -50,21 +50,16 @@ void Gizmos::drawLine(const std::string& id, glm::vec3 from, glm::vec3 to, float
     }
 }
 
+void Gizmos::drawLine(const std::string& id, glm::vec3 from, glm::vec3 to, float lifespan, Space space)
+{
+    auto gizmo = std::make_shared<LineGizmo>(id, from, to, mColor, lifespan);
+    push(gizmo, space);
```
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/gizmos.cpp:56]:
```suggestion
    push(std::make_shared<LineGizmo>(id, from, to, mColor, lifespan), space);
```

File: engine/src/cubos/engine/gizmos/gizmos.cpp [hunk 4]
```diff
@@ -50,21 +50,16 @@ void Gizmos::drawLine(const std::string& id, glm::vec3 from, glm::vec3 to, float
...
+}
+
 void Gizmos::drawBox(const std::string& id, glm::vec3 corner, glm::vec3 oppositeCorner, float lifespan, Space space)
 {
     auto gizmo = std::make_shared<BoxGizmo>(id, corner, oppositeCorner, mColor, lifespan);
```
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/gizmos.cpp:61]:
```suggestion
```

File: engine/src/cubos/engine/gizmos/gizmos.cpp [hunk 5]
```diff
@@ -50,21 +50,16 @@ void Gizmos::drawLine(const std::string& id, glm::vec3 from, glm::vec3 to, float
...
-    }
-    else if (space == Space::View)
-    {
-        viewGizmos.push_back(gizmo);
-    }
-    else if (space == Space::Screen)
-    {
-        screenGizmos.push_back(gizmo);
-    }
+    push(gizmo, space);
```
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/gizmos.cpp:62]:
```suggestion
    push(std::make_shared<BoxGizmo>(id, corner, oppositeCorner, mColor, lifespan), space);
```

File: engine/src/cubos/engine/gizmos/gizmos.cpp [hunk 6]
```diff
@@ -93,3 +88,11 @@ void Gizmos::drawWireBox(const std::string& id, glm::vec3 corner, glm::vec3 oppo
              glm::vec3{oppositeCorner[0], corner[1], oppositeCorner[2]}, lifespan, space);
     drawLine(id, glm::vec3{oppositeCorner[0], oppositeCorner[1], corner[2]}, oppositeCorner, lifespan, space);
 }
+
+void Gizmos::drawCutCone(const std::string& id, glm::vec3 firstBaseCenter, float firstBaseRadius,
+                         glm::vec3 secondBaseCenter, float secondBaseRadius, float lifespan, Space space)
+{
+    auto gizmo = std::make_shared<CutConeGizmo>(id, firstBaseCenter, firstBaseRadius, secondBaseCenter,
+                                                secondBaseRadius, mColor, lifespan);
+    push(gizmo, space);
```
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/gizmos.cpp:97]:
```suggestion
    push(std::make_shared<CutConeGizmo>(id, firstBaseCenter, firstBaseRadius, secondBaseCenter,
                                                secondBaseRadius, mColor, lifespan), space);
```

File: engine/src/cubos/engine/gizmos/renderer.hpp [hunk 7]
```diff
@@ -20,15 +22,17 @@ namespace cubos::engine
...
-        Primitive linePrimitive; ///< GL line information
-        Primitive boxPrimitive;  ///< GL box information
+        Primitive linePrimitive;    ///< GL line information.
+        Primitive boxPrimitive;     ///< GL box information.
+        Primitive cutConePrimitive; ///< GL cut cone information.
 
-        /// @brief Sets up the GizmosRenderer to be used
-        /// @param renderDevice the current RenderDevice being used
+        /// @brief Sets up the GizmosRenderer to be used.
+        /// @param renderDevice the current RenderDevice being used.
```
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/renderer.hpp:30]:
```suggestion
        /// @brief Sets up the render device to be used.
        /// @param renderDevice Render device.
```

File: engine/src/cubos/engine/gizmos/types/box.hpp [hunk 8]
```diff
@@ -8,19 +8,19 @@
...
-        /// @param corner point at one of the corners of the box
-        /// @param oppositeCorner point at the opposite corner of the box
-        /// @param color color for the gizmo to be drawn in
-        /// @param lifespan time the gizmo will remain on screen, in seconds
+        /// @brief BoxGizmo constructor.
+        /// @param id identifier of the gizmo.
+        /// @param corner point at one of the corners of the box.
+        /// @param oppositeCorner point at the opposite corner of the box.
+        /// @param color color for the gizmo to be drawn in.
+        /// @param lifespan time the gizmo will remain on screen, in seconds.
```
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/types/box.hpp:23]:
```suggestion
        /// @param id Identifier of the gizmo.
        /// @param corner Point at one of the corners of the box.
        /// @param oppositeCorner Point at the opposite corner of the box.
        /// @param color Color for the gizmo to be drawn in.
        /// @param lifespan Time the gizmo will remain on screen, in seconds.
```

File: engine/src/cubos/engine/gizmos/types/box.hpp [hunk 9]
```diff
@@ -29,8 +29,9 @@ namespace cubos::engine
         {
         }
 
-        /// @brief draws the gizmo to screen
-        /// @param renderer the GizmosRenderer in use
+        /// @brief draws the gizmo to screen.
+        /// @param renderer the GizmosRenderer in use.
+        /// @param mvp matrix containing projection and viewpoint transformations.
```
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/types/box.hpp:34]:
```suggestion
        /// @brief Draws the gizmo to screen.
        /// @param renderer GizmosRenderer in use.
        /// @param mvp Matrix containing projection and viewpoint transformations.
```

File: engine/src/cubos/engine/gizmos/types/cut_cone.hpp [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+#include <glm/gtc/matrix_transform.hpp>
+#include <glm/gtx/rotate_vector.hpp>
+
+#include <cubos/engine/gizmos/gizmos.hpp>
+
+#include "../renderer.hpp"
+
+namespace cubos::engine
+{
+    /// @brief A gizmo that is a cylinder with bases that can be different.
```
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/types/cut_cone.hpp:12]:
```suggestion
    /// @brief A cylinder gizmo with bases that can be different.
```

File: engine/src/cubos/engine/gizmos/types/cut_cone.hpp [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    class CutConeGizmo : public cubos::engine::Gizmos::Gizmo
```
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/types/cut_cone.hpp:13]:
```suggestion
    class CutConeGizmo : public Gizmos::Gizmo
```

File: engine/src/cubos/engine/gizmos/types/cut_cone.hpp [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+
+    public:
+        /// @brief BoxGizmo constructor.
+        /// @param id identifier of the gizmo.
+        /// @param firstBaseCenter center of one of the bases.
+        /// @param firstBaseRadius radius of one of the bases.
+        /// @param secondBaseCenter center of the second base.
+        /// @param secondBaseRadius radius of the second base.
+        /// @param color color for the gizmo to be drawn in.
+        /// @param lifespan time the gizmo will remain on screen, in seconds.
```
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/types/cut_cone.hpp:28]:
```suggestion
        /// @brief Constructs.
        /// @param id Identifier of the gizmo.
        /// @param firstBaseCenter Center of one of the bases.
        /// @param firstBaseRadius Radius of one of the bases.
        /// @param secondBaseCenter Center of the second base.
        /// @param secondBaseRadius Radius of the second base.
        /// @param color Color for the gizmo to be drawn in.
        /// @param lifespan Time the gizmo will remain on screen, in seconds.
```

File: engine/src/cubos/engine/gizmos/types/cut_cone.hpp [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+            , mPointA(firstBaseCenter)
+            , mRadiusA(firstBaseRadius)
+            , mPointB(secondBaseCenter)
+            , mRadiusB(secondBaseRadius)
+        {
+        }
+
+        /// @brief draws the gizmo to screen.
+        /// @param renderer the GizmosRenderer in use.
+        /// @param mvp matrix containing projection and viewpoint transformations.
```
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/types/cut_cone.hpp:41]:
```suggestion
        /// @brief Draws the gizmo to the screen.
        /// @param renderer GizmosRenderer in use.
        /// @param mvp Matrix containing projection and viewpoint transformations.
```

File: engine/src/cubos/engine/gizmos/types/cut_cone.hpp [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+        void draw(cubos::engine::GizmosRenderer& renderer, const glm::mat<4, 4, float, glm::packed_highp>& mvp) override
```
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/types/cut_cone.hpp:42]:
```suggestion
        void draw(GizmosRenderer& renderer, const glm::mat<4, 4, float, glm::packed_highp>& mvp) override
```

File: engine/src/cubos/engine/gizmos/types/cut_cone.hpp [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+        {
+            auto* verts = static_cast<glm::vec3*>(renderer.cutConePrimitive.vb->map());
+
+            glm::vec3 n = glm::normalize(mPointB - mPointA);
+            glm::vec3 p = {n[1], -n[0], n[2]};
+
+            glm::vec3 pA = p * mRadiusA;
+
+            for (int i = 0; i < cubos::engine::GizmosRenderer::CutConeVertsPerBase; i++)
```
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/types/cut_cone.hpp:51]:
```suggestion
            for (int i = 0; i < GizmosRenderer::CutConeVertsPerBase; i++)
```

File: engine/src/cubos/engine/gizmos/types/cut_cone.hpp [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+            {
+                glm::vec3 vert =
+                    glm::rotate(
+                        pA, (float)i * glm::radians(360.0F / (float)cubos::engine::GizmosRenderer::CutConeVertsPerBase),
```
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/types/cut_cone.hpp:55]:
```suggestion
                        pA, (float)i * glm::radians(360.0F / (float)GizmosRenderer::CutConeVertsPerBase),
```

File: engine/src/cubos/engine/gizmos/types/cut_cone.hpp [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+                        n) +
+                    mPointA;
+                verts[i] = {vert[0], vert[1], vert[2]};
+            }
+
+            glm::vec3 pB = p * mRadiusB;
+
+            for (int i = cubos::engine::GizmosRenderer::CutConeVertsPerBase;
+                 i < 2 * cubos::engine::GizmosRenderer::CutConeVertsPerBase; i++)
```
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/types/cut_cone.hpp:64]:
```suggestion
            for (int i = GizmosRenderer::CutConeVertsPerBase;
                 i < 2 * GizmosRenderer::CutConeVertsPerBase; i++)
```

File: engine/src/cubos/engine/gizmos/types/cut_cone.hpp [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+            {
+                glm::vec3 vert =
+                    glm::rotate(
+                        pB, (float)i * glm::radians(360.0F / (float)cubos::engine::GizmosRenderer::CutConeVertsPerBase),
```
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/types/cut_cone.hpp:68]:
```suggestion
                        pB, (float)i * glm::radians(360.0F / (float)GizmosRenderer::CutConeVertsPerBase),
```

File: engine/src/cubos/engine/gizmos/types/line.hpp [hunk 19]
```diff
@@ -8,28 +8,29 @@
...
-        /// @param from point at one of the ends of the line
-        /// @param to point at the other end of the line
-        /// @param color color for the gizmo to be drawn in
-        /// @param lifespan time the gizmo will remain on screen, in seconds
+        /// @brief LineGizmo constructor.
+        /// @param id identifier of the gizmo.
+        /// @param from point at one of the ends of the line.
+        /// @param to point at the other end of the line.
+        /// @param color color for the gizmo to be drawn in.
+        /// @param lifespan time the gizmo will remain on screen, in seconds.
```
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/types/line.hpp:23]:
```suggestion
        /// @brief Constructs.
        /// @param id Identifier of the gizmo.
        /// @param from Point at one of the ends of the line.
        /// @param to Point at the other end of the line.
        /// @param color Color for the gizmo to be drawn in.
        /// @param lifespan Time the gizmo will remain on screen, in seconds.
```

File: engine/src/cubos/engine/gizmos/types/line.hpp [hunk 20]
```diff
@@ -8,28 +8,29 @@
...
             , mPointA(from)
             , mPointB(to)
         {
         }
 
-        /// @brief draws the gizmo to screen
-        /// @param renderer the GizmosRenderer in use
+        /// @brief draws the gizmo to screen.
+        /// @param renderer the GizmosRenderer in use.
+        /// @param mvp matrix containing projection and viewpoint transformations.
```
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/types/line.hpp:33]:
```suggestion
        /// @brief Draws the gizmo to screen.
        /// @param renderer GizmosRenderer in use.
        /// @param mvp Matrix containing projection and viewpoint transformations.
```

**Review by RiscadoA — CHANGES_REQUESTED** [review]
Thanks for working on this! The logic is ok, just some minor nitpicks

**Review by RiscadoA — APPROVED** [review]

**Pull request merged**